### PR TITLE
Populate required field Message with n/a if it is empty

### PR DESCRIPTION
**Pull request by NeilMountford**
Fix Issue #1066.

## Changes
Application Insights rejects exceptions that have empty messages. This means that some exceptions are lost, making investigation of runtime issues difficult. I noticed this when investigating an issue in a .NET 8 API, where a `CosmosException` for a conflict occurred, was logged on the console but wasn't logged in App Insights.

There was previously a PR #2697 that never got merged due to lack of tests.

Original example from the other PR:
```csharp
var configuration = TelemetryConfiguration.CreateDefault();
configuration.ConnectionString = "InstrumentationKey=...";
var telemetryClient = new TelemetryClient(configuration);
telemetryClient.TrackException(new Exception("")); // ignored
telemetryClient.TrackException(new Exception("foo")); // logged
telemetryClient.Flush();
```

Original response from the other PR:
```json
{
    "itemsReceived": 1,
    "itemsAccepted": 0,
    "errors":
    [
        {
            "index": 0,
            "statusCode": 400,
            "message": "109: Field 'message' on type 'ExceptionDetails' is required but missing or empty. Expected: string, Actual: undefined"
        }
    ]
}
```

### Checklist
- [✅] I ran Unit Tests locally.*
- [✅] CHANGELOG.md updated with one line description of the fix, and a link to the original issue if available.

* I had some test failures on a fresh clone, but the tests for the relevant part of the code were passing, so I'm hoping it's just an issue with some dependencies on my machine.

For significant contributions please make sure you have completed the following items:

- [N/A] Design discussion issue #
- [N/A] Changes in public surface reviewed

The PR will trigger build, unit tests, and functional tests automatically. Please follow [these](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/.github/CONTRIBUTING.md) instructions to build and test locally.

### Notes for authors:
- FxCop and other analyzers will fail the build. To see these errors yourself, compile localy using the Release configuration.

### CODE REVIEW

**Comment by NeilMountford** [comment]
@microsoft-github-policy-service agree

**Review by cijothomas — APPROVED** [review]
Thanks, this looks good!

Will ask one more owner to review before merge, as we are adding limited additions (including fixes) to this repo.

**Comment by NeilMountford** [comment]
Thanks @cijothomas, I've pushed a further change to move `using Microsoft.ApplicationInsights.DataContracts;` into the namespace block to make the build happy for the net4* builds - apologies for not catching this locally, I don't have access to a Windows machine currently. Could you approve the workflows again please? 🙂

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/ExceptionConverterTest.cs [hunk 1]
```diff
@@ -185,6 +185,19 @@ public void DoesNotTrimShortExceptionMessages()
             Assert.AreEqual(5, expDetails.message.Length);
         }
 
+        [TestMethod]
+        [DataRow(null)]
+        [DataRow("")]
+        [DataRow(" ")]
+        public void ProvidesDefaultMessage(string message)
```
Comment (TimothyMothra) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/ExceptionConverterTest.cs:192]: What's the intention of this parameter? I don't see it used.

Comment (rajkumar-rangaraj) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/ExceptionConverterTest.cs:192]: `message` parameter is not used within this method.

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/External/ExceptionDetailsImplementation.cs [hunk 2]
```diff
@@ -21,7 +22,7 @@ internal static ExceptionDetails CreateWithoutStackInfo(Exception exception, Exc
             {
                 id = exception.GetHashCode(),
                 typeName = exception.GetType().FullName,
-                message = exception.Message,
+                message = Utils.PopulateRequiredStringValue(exception.Message, "message", typeof(ExceptionTelemetry).FullName),
```
Comment (TimothyMothra) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/External/ExceptionDetailsImplementation.cs:25]: nit: note that this helper method is using `string.IsNullOrEmpty`. 
https://github.com/microsoft/ApplicationInsights-dotnet/blob/bd8b0c25712e6f3c5275ae06b0f0e9331340fee4/BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Utils.cs#L44-L49

I see in your test cases you passed in a string with a space `" "`. Note that this helper method won't catch this.

**Review by TimothyMothra — CHANGES_REQUESTED** [review]
please address comments

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/ExceptionConverterTest.cs (hunk 1 above)
Comment (TimothyMothra) [reply BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/ExceptionConverterTest.cs:192]: my comment is a dupe of https://github.com/microsoft/ApplicationInsights-dotnet/pull/2857#discussion_r1526645511

Comment (NeilMountford) [reply BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/ExceptionConverterTest.cs:192]: My bad, started with just an empty string test then decided to add `DataRow`s to cover null and whitespace too. I've fixed the test to use the parameter now.

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/External/ExceptionDetailsImplementation.cs (hunk 2 above)
Comment (NeilMountford) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/External/ExceptionDetailsImplementation.cs]: I fixed the test to use the parameter and it failed for the null and whitespace examples. 

- The null failure was due to the `Message` property on `Exception` returning a default if the `_message` field is null. I've added a custom exception for the test that overrides Message and returns the provided message as is.
- The empty string failure was down to the `.IsNullOrEmpty` so I've added a new, similar helper method that instead uses the provided `IsNullOrWhitespace` extension. In cases where it is null/whitespace I've used the same logging method, but can add another slightly different one with a different EventId if you prefer?

**Review by TimothyMothra — APPROVED** [review]
LGTM

**Comment by NeilMountford** [comment]
@TimothyMothra What are the next steps to get this merged/released? Let me know if you need anything from me 🙂

**Comment by jiilaa** [comment]
So, half a year since this fix was merged (and even longer since the latest release for this library), maybe a good time for a release? :)

**Comment by asos-AdamCox** [comment]
@TimothyMothra / @cijothomas are chance of a release? We've just spent an entire day chasing this bug down only to find it was found in 2019, fixed 7 months ago and still not released...

**Pull request merged**